### ISSUE TITLE
How to one-hot encode a multi-class dataset and how to use Boundary Loss on B x N x W x H logits?

### ISSUE THREAD

**Issue by sarmientoj24** [LIVIAETS/boundary-loss#56]
Basically two questions,

(1) If I have the following dataset target, of size B x W x H, how do I encode it for your distance map?

```
[
  [0 0 0 0 1 1 1 2 1],
  [0 0 0 0 1 1 1 2 1],
  [0 0 1 1 2 1 0 0 1],
  [0 0 2 2 2 1 0 0 1],
]
```

and (2), how do I use a B x N x W x H in **logits** into your Boundary Loss?

Extra Question: Does this have an asymmetrical boundary loss version?

**Comment by dujunl** [comment]
Hello, I would like to ask if you have solved the problem of multi-classification?

**Comment by HKervadec** [comment]
(Sorry for the _very_ delayed reply. Somehow the Github notification for this repo are still broken on my side.)

(1) you would chain the [`class2one_hot`](https://github.com/LIVIAETS/boundary-loss/blob/master/utils.py#L227) to the [`one_hot2dist`](https://github.com/LIVIAETS/boundary-loss/blob/master/utils.py#L245) functions. Notice that the `distmap_transform` of the dataloader already do that https://github.com/LIVIAETS/boundary-loss/blob/master/dataloader.py#L105, calling first the `gt_transform` before doing the distance transform. All that needs to be set correctly, is the parameter `K` (for number of classes). You should get an error is `K` is set too low for your number of classes in the ground truth.

(2) if you look at the boundary loss (which is just element-wise multiplication) you notice this is already the case: https://github.com/LIVIAETS/boundary-loss/blob/master/losses.py#L89: 
```python
pc = probs[:, self.idc, ...].type(torch.float32)
dc = dist_maps[:, self.idc, ...].type(torch.float32)

multipled = einsum("bkwh,bkwh->bkwh", pc, dc)

loss = multipled.mean()
```
I use [`einsum`](https://pytorch.org/docs/stable/generated/torch.einsum.html) to label the axes explicitly. Here the only different is that I call your `N` axis `k`.

Hope this helps!
